Cleanup several aspects of tech debt around graphics and font management

mike-spa opened the pull request:
Resolves: #30456
Resolves: #14435
Resolves: #12714

Please refer to the commit descriptions for explanations about the specific changes. Should be reviewed one commit at a time so that what I'm doing makes more sense.

Here is a summary of the main changes:

- All of our font metrics functions now use Qt. Before, we were getting some metrics from Qt and others from our own implementation based on Freetype.
- The font metrics functions have been streamlined and reorganized. The distinction betwen char16 and char32 types has been removed. The special symBbox and symAdvance methods have been removed.
- The constants associated to graphics resolution in mscore.h have been cleaned up, namely DPI_F has been removed, SPATIUM20 has been removed, MScore::pixelRatio has been removed, DPI has been increased to 1200. 
- Our default spatium is now properly defined as 1.75 mm * DPMM (as opposed to the previous magic number of 24.8)
- The default Smufl font size is set to the precise value corresponding to 7mm staves, as opposed to being approximately set to 20 pt and requiring subsequent correction (that's what SPATIUM20 was needed for).
- Draw-related methods have been removed from the TextBase class.
- The bounding boxes of our music symbols are now calculated with true vectorial precision. This avoids the approximations due to rasterization and guarantees perfect consistency across platforms. 


All the vtests will fail by rounding errors.

mike-spa committed:
Remove DPI_F

I have never really understood what this quantity represents, other than it has a value of 5 and was multiplied by 72 to obtain the DPI.

From what I can gather, it seems that the idea is to start from 72 points per inch (which comes from the definition of the typographical point, defined as 1/72 of an inch) and to set resolution of the program to 5 times that. In that sense DPI_F is probably intended as a "fractional" resolution with respect to the typographical point. That results in a DPI of 360.

It now seems that this DPI_F variable is effectively unused (all the functions that take it as parameter end up ignoring it). And the DPI can be just set to 360, I don't see why it has to be constructed as an integer multiple of a typographical unit (in fact most printers have resolutions of 600 or 1200 DPI, which are not multiples of 72).

mike-spa committed:
Remove use of FreeType from QFontProvider

The reason why that code exist is that we handle text via Qt font metrics but music symbols via FreeType metrics. That is now managed by FontProviderDispatcher so there is no reason to keep this code hidden behind a compiler flag. In fact it doesn't even compile if the flag is removed.

mike-spa committed:
Bypass the FontProviderDispatcher and use only Qt for everything

The necessity of using Qt for some font metrics and FreeType for others came from #22189, which was caused by a Qt bug. Qt has now resolved that bug by introducing the flag PreferTypoLineMetrics. So there is no reason now for using two different source of metrics for text. In fact, I see also no reason to not use Qt for the music symbols fonts. I don't know the historical reason for that, but Qt should be able to handle them like any other font. To me this just feels like historical tech debt.

Additionally, given that all of our drawing uses QPainter, using Qt for the metrics that define such drawing feels more logical to me and more likely to work consistently in all situations.

Incidentally, this also happens to fix #30456. I don't know why, but FreeType returns wrong values for the line spacing, while Qt seems to give the correct values.

mike-spa committed:
Increase resolution to 1200 DPI

After switching the symbol metrics to Qt, I found that Qt wasn't returning very precise bounding boxes. They seemed to all have some (small, but noticeable) rounding errors. By contrast, the bounding boxes that we were getting before were pixel-perfect. This may well be the historical reason why Qt metrics were not used for music symbol.

It turns out though that Qt wasn't imprecise, it was just limited by the resolution of 360 DPI that we were setting ourselves. I assume that QFontMetricsF, despite returning double precision values, internally still rounds to whatever resolution is set. Increasing the resolution to 1200 DPI makes Qt's metrics on music symbols also pixel-perfect.

As far as I understand, this resolution value only affects internal calculations, not the actual rendering resolution (plus music font metrics are cached anyway), so I don't think this should have any impact on performance.

This also requires updating the value of default spatium (having finally understood what that 24.8 was). That number implicitely assumed a DPI of 360. We should obviously just write the spatium with the proper conversion factor from millimetres to pixels/points. Luckily this doesn't affect existing files, where it's already written in millimitres.

mike-spa committed:
Fix insane hack for brace drawing

I have no idea what that magic number of 100 was about nor why it was even working before the change of resolution. It definitely wasn't working anymore with the new resolution.

mike-spa committed:
Remove symBBox and symAdvance methods from FontProvider

Music symbols are just characters in a font, so there is no reason why their metrics should need special functions.

mike-spa committed:
Fix QFontProvider::inFontUcs4

mike-spa committed:
Fix error in accidental layout

Error was always there but somehow was made visible by the change of resolution.

mike-spa committed:
Remove SPATIUM20 variable

This is another variable that has never made any sense to me whatsoever. I can only infer from the way it was used that it was meant to represent the default value of spatium in pixel units. But if that's the case, the way in which it was defined made no sense at all. DPI / 72 used to just be equal to DPI_F, so SPATIUM20 was equal to 5 * DPI_F = 25. To me, the only reason why this seemed to work correctly as the default spatium value is that it just happened to be close to 24.8, which was the actual value.

[Spoiler: see commit 577929d41b1416e1918dd9fa8159b091826f8598 for the explanation of what SPATIUM_20 meant]

This commit removes SPATIUM20 and replaces it with appropriate calls to the style defaults. In a couple of places (namely StaffType and EngravingFont) the Score and the corresponding Style haven't been created yet, so they need to access the StyleDef class directly. In fact I don't see what StyleDef should make its members private. I think they can be public cause they are const so they can't be changed.

mike-spa committed:
Remove drawTextWorkaround

No idea about the exact history of this. It seems to have been introduced as a Mac-specific hack ages ago to solve some Qt-related problem. Pretty sure it's not necessary anymore (also produces no difference in vtests).

mike-spa committed:
Remove draw method from TextBase class

I think this was the only DOM class that still had its own draw method. I have moved it to the respective draw module where it belongs.

cbjeukendrup commented on src/framework/draw/internal/fontprovider.cpp:
```diff
@@ -105,14 +105,3 @@ RectF FontProvider::tightBoundingRect(const muse::draw::Font& f, const muse::Str
-// Score symbols
-RectF FontProvider::symBBox(const muse::draw::Font& f, char32_t ucs4) const
-{
-    return fontsEngine()->symBBox(f, ucs4);
```
Remove the methods from `FontsEngine` as well?

cbjeukendrup commented on src/engraving/internal/engravingfont.cpp:
```diff
@@ -852,8 +852,8 @@ void EngravingFont::computeMetrics(EngravingFont::Sym& sym, const Smufl::Code& c
-        sym.bbox = fontProvider()->symBBox(m_font, sym.code);
-        sym.advance = fontProvider()->symAdvance(m_font, sym.code);
+        sym.bbox = fontProvider()->boundingRect(m_font, sym.code);
+        sym.advance = fontProvider()->horizontalAdvance(m_font, sym.code);
```
_Maybe_ (part of) the reason was that that `sym.code` doesn't fit in `Char` for all symbols but needs `char32_t` instead?

mike-spa replied:
So, about the first point: I ended up generalizing all the metric functions to 32bit characters, cause there's really no reason not to. That way we can also get rid of the inFont vs inFontUcs4 distinction.

About the second point: yes the Smufl code points are all within the 16 bit range, however I discovered that the unicode standard also has a range dedicated to music symbols which is beyond the 16 bit range. Smufl suggests that if a symbol exists both in Smufl and in Unicode it should be included at both code points but some of our fonts only included it at the Unicode point, so we do need to keep the 32bit variable after all.

cbjeukendrup commented on src/framework/draw/internal/qfontprovider.cpp:
```diff
@@ -25,6 +25,7 @@
 #include <QPaintDevice>
 #include <QFontDatabase>
 #include <QFontMetricsF>
+#include <qrawfont.h>
```
```suggestion
#include <QRawFont>
```

cbjeukendrup commented on src/engraving/rendering/score/accidentalslayout.h:
```diff
@@ -76,6 +76,9 @@ class AccidentalsLayout
         size_t smallGroupLimit() const { return m_smallGroupLimit; }
 
+        bool updateColumns() const { return m_updateColumns; }
+        void setUpdateColumns(bool v) { m_updateColumns = v; }
```
Unused?

cbjeukendrup reviewed (commented):
Great work! A few minor comments.

cbjeukendrup commented:
> removing MScore::pixelRatio

I think this should go hand in hand with removing `NotationConfiguration::notationScaling()` / `scalingFromZoomPercentage`.
Instead, the active DPI should come from the `QPainter`'s [`QPaintDevice`](https://doc.qt.io/qt-6/qpaintdevice.html). 
There is some scaling needed to compensate for the fact that all values in the engraving module assume 1200 DPI, while the QPainter's DPI is different.
That compensation is done by scaling by the value returned by `scalingFromZoomPercentage`. 
But then for drawing text, my understanding is that we basically need to undo that compensation, because all text sizes are expressed in point size, so are already relative to the QPainter's DPI. 

As a solution, I would propose:
* remove `MScore::pixelRatio` and those `NotationConfiguration` methods
* take the DPI value from the QPainter, and perform the DPI compensation scaling in `Paint::paintScore`, not in `AbstractNotationPaintView` etc
* pass the compensation scaling factor to the individual draw functions via `rendering::PaintOptions`, so that it can be temporarily undone whenever some text needs to be painted

cbjeukendrup commented:
Another option might be to figure out what's the difference between QFont::pixelSize and QFont::pointSize, and whether that can somehow help us

miiizen committed:
Fix chord symbol compatibility

mike-spa committed:
Generalize all font metrics function to char32_t

Cause that will work with both 16bit and 32bit (i.e. ucs4) character maps

mike-spa committed:
Code review fixes and further cleanup

mike-spa committed:
Remove MScore::pixelRatio

MuseScore's internal coordinates are expressed in units determined by mu::engraving::DPI. The only quantity that is not expressed in this unit is text size, which is expressed in points and ignores the DPI. In our score model, doubling the DPI means for example that point (10, 10) becomes (20, 20), but a text of 10pt size does not become 20pt, it remains 10pt.

When we draw on screen we apply a transform that maps our coordinates to the screen's physical coordinates, but because the text size is expressed in different units we need to apply a different scaling factor. On top of that, QFont::pointSizeF is also DPI-aware, meaning QPainter will apply its own scaling factor based on the painter's DPI to determine the resulting pixel size.

QFont::setPixelSize is not a viable solution because it only accepts integer values. Another solution would be applying an equivalent scaling transform to the painter, but that results awkward because we need to scale only the text size but not the position. In the end the better solution is using QFont::setPointSizeF. What is bad about the current implementation is that the entire program has to be aware of the painter's DPI via MScore::pixelRatio. This is not necessary because any painter implementation is aware of its own logical DPI, so the scaling factor can be applied in the Painter class and no other part of code needs to know about it.

mike-spa committed:
Set internal unit to 1/100 of millimetre

In theory, this unit is entirely arbitrary, so we could also set it to 1mm (that would make debugging nicer, cause every number we see represents millimetres). However, that corresponds to a very low DPI value (25.4), which doesn't work nicely with QFontMetrics, and very small scaled point sizes when drawing, which doesn't work nicely with QPainter. This seems the better compromise.

mike-spa committed:
Fix utests

cbjeukendrup commented on src/framework/draw/painter.cpp:
```diff
@@ -420,8 +427,18 @@ void Painter::drawRoundedRect(const RectF& rect, double xRadius, double yRadius)
+void Painter::applyFontSizeScaling()
+{
+    Font f = font();
+    double scaledPointSize = f.pointSizeF() * mu::engraving::DPI / deviceLogicalDpi();
```
I'm afraid we can't use `mu::engraving` in `framework`

cbjeukendrup commented on src/importexport/musicxml/internal/musicxml/import/importmusicxmlpass1.cpp:
These variables are now unused

cbjeukendrup commented on src/framework/draw/bufferedpaintprovider.cpp:
```diff
@@ -256,6 +256,11 @@ void BufferedPaintProvider::restore()
 
+double BufferedPaintProvider::deviceLogicalDpi() const
+{
+    return 0;
```
Given that Painter divides by this value, `0` might not be the best dummy value (even while this class is currently not really used)

mike-spa committed:
Code review fixes

igorkorsukov commented on src/engraving/dom/stafftype.cpp:
```diff
@@ -522,6 +522,11 @@ void StaffType::setDurationFontName(const String& name)
 
+double StaffType::defaultSpatium() const
+{
+    return StyleDef::styleValues[static_cast<size_t>(Sid::spatium)].defaultValue().toDouble();
```
Maybe it would be better to return the `Spatium` type here?

igorkorsukov replied:
Hmm, then we need a different name, otherwise it's very confusing now.

mike-spa committed:
Fix default smufl font size

It has somehow become common knowledge that the Smufl font size corresponding to a 7 mm staff (i.e. a spatium of 1.75 mm) is 20 pt. I too thought that that was a definition. Turns out it isn't. Smufl just prescribes that the glyphs are scaled in a way that the em size of the font corresponds to a 5-line staff height, and it just so happens that 7 mm are close to 20 typographical points (7mm / 25.4 * 72 = 19.84 pt).

Instead of computing all the font metrics using the correct point size of 19.84 pt, corresponding to our 7mm default, the metrics were being computed at 20 pt, and the SPATIUM_20 constant (which means "the value of spatium corresponding to 20 pt", different from the actual spatium) was being used throughout the codebase as a correction factor to account for the difference between 19.84 and 20.0, most notably in EngravingItem::magS. In other words we were scaling down the size of all items by a correction factor instead of computing it correctly from the start.

Having removed SPATIUM_20, it was now necessary to compute the Smufl font metrics at the correct point size.

mike-spa committed:
Revert resolution back to 1200 DPI

Turns out that the resolution does have a performance impact on QFontMetrics, but it's only noticeable in pieces with tons of text (typically lyrics) because unlike music symbol fonts those values can't be cached. It also seems to only start scaling badly for values beyond 1200 DPI, while going below seems to make no difference, so we can settle on 1200 DPI.

mike-spa committed:
Use real vector calculations for boundingRect of music glyphs

mike-spa committed:
Update vtests

mike-spa committed:
Style::defaultSpatium optimazation

Having eliminated SPATIUM_20 and replaced it with calls to defaultSpatium(), this function gets called a lot now.

No need to check old defaults: the default spatium was always 1.75mm.

Saves around 3% layout time on my test file.
